cns/deviceplugin/server.go

```diff
@@ -0,0 +1,153 @@
+package deviceplugin
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+type deviceCounter interface {
+	getDeviceCount() int
+}
+
+type Server struct {
+	address              string
+	logger               *zap.Logger
+	deviceCounter        deviceCounter
+	listAndWatchDoneChan chan struct{}
+	deviceCheckInterval  time.Duration
+}
+
+func NewServer(address string, logger *zap.Logger, deviceCounter deviceCounter, deviceCheckInterval time.Duration) *Server {
+	return &Server{
+		address:              address,
+		logger:               logger,
+		deviceCounter:        deviceCounter,
+		listAndWatchDoneChan: make(chan struct{}),
+		deviceCheckInterval:  deviceCheckInterval,
+	}
+}
+
+// Run starts the grpc server and blocks until an error or context is cancelled. Wait on Ready to know when the server is ready.
+func (s *Server) Run(ctx context.Context) error {
+	grpcServer := grpc.NewServer()
+	v1beta1.RegisterDevicePluginServer(grpcServer, s)
+	defer close(s.listAndWatchDoneChan)
+
+	l, err := net.Listen("unix", s.address)
+	if err != nil {
+		return errors.Wrap(err, "error listening on socket")
+	}
+	errChan := make(chan error, 2) //nolint:gomnd // disabled in favor of readability
+	go func() {
+		errChan <- grpcServer.Serve(l)
+	}()
+	defer grpcServer.Stop()
+
+	select {
+	case err := <-errChan:
+		return errors.Wrap(err, "error running grpc server")
+	case <-ctx.Done():
+		return nil
+	}
```

Comment:
Updated code. In addition to what is suggested above,
1) Added defer l.close() to prevent resource leak
2) Used GracefulStop() instead of Stop()
3) Added error handing for  grpcServer.Serve(l). It is needed by linter as well